File: 125_algorithms/_exercises/templates/_algorithms_challenges/leetcode/leetCode/DynamicProgramming/53_MaximumSubarray.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-


c.. Solution o..
    # O(n) space
    ___ maxSubArray  nums
        num_len = l..(nums)
        # dp[i]: Largest sum of contiguous subarray start from i
        dp = [-1] * num_len
        max_sum = dp[num_len - 1] = nums[num_len - 1]

        ___ i __ r..(num_len - 2, -1, -1
            dp[i] = m..(nums[i], dp[i+1]+nums[i])
            max_sum = m..(dp[i], max_sum)

        r_ max_sum


c.. Solution_2 o..
    # DP same with the previous, but O(1) space
    ___ maxSubArray  nums
        num_len = l..(nums)
        max_sum = pre_sum = nums[num_len - 1]

        ___ i __ r..(num_len - 2, -1, -1
            pre_sum = m..(nums[i], pre_sum+nums[i])
            max_sum = m..(pre_sum, max_sum)

        r_ max_sum

"""
[-1]
[1]
[-9,-2,-3,-5,-3]
[-2,1,-3,4,-1,2,1,-5,4]
"""

File: 125_algorithms/_exercises/templates/_algorithms_challenges/leetcode/leetCode/DynamicProgramming/53_MaximumSubarray.cpp
/*
 * @Author: xuezaigds@gmail.com
 * @Last Modified time: 2016-04-07 11:08:44
 */


c.. Solution {
public:
    int maxSubArray(vector<int>& nums) {
        int nums_len = nums.size();
        int pre_sum = nums[0];
        int max_sum = nums[0];

        ___(int i=1;i<nums_len;i++){
            pre_sum = m..(pre_sum+nums[i], nums[i]);
            max_sum = m..(max_sum, pre_sum);
        }
        r_ max_sum;
    }
};

/*
[-1]
[1]
[-9,-2,-3,-5,-3]
[-2,1,-3,4,-1,2,1,-5,4]
*/
